### PR TITLE
feat(env-check): remove dependency of system package manager from node env checker

### PR DESCRIPTION
**Pull request by COLDTURNIP**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#10019

#### What this PR does / why we need it:

Instead of reading the system package database, check the executables needed by the system, to eliminate the system management tool chain issue.

### CODE REVIEW

File: controller/monitor/environment_check_monitor.go [hunk 1]
```diff
@@ -203,21 +201,25 @@ func (m *EnvironmentCheckMonitor) syncPackagesInstalled(kubeNode *corev1.Node, n
...
+		packageProbeExecutables["cryptsetup"] = "cryptsetup"
+		packageProbeExecutables["device-mapper"] = "dmsetup"
 	case strings.Contains(osImage, "gentoo"):
-		queryPackagesCmd = "qlist"
-		options = append(options, "-I")
-		packages = append(packages, "net-fs/nfs-utils", "sys-block/open-iscsi", "sys-fs/cryptsetup", "sys-fs/lvm2")
+		packageProbeExecutables["net-fs/nfs-utils"] = "mount.nfs"
+		packageProbeExecutables["sys-block/open-iscsi"] = "iscsiadm"
+		packageProbeExecutables["sys-fs/cryptsetup"] = "cryptsetup"
+		packageProbeExecutables["sys-fs/lvm2"] = "dmsetup"
```
Comment (derekbit) [inline controller/monitor/environment_check_monitor.go:222]: Excellent!

File: controller/monitor/environment_check_monitor.go [hunk 2]
```diff
@@ -190,10 +188,10 @@ func (m *EnvironmentCheckMonitor) syncPackagesInstalled(kubeNode *corev1.Node, n
 	case strings.Contains(osImage, "ubuntu"):
 		fallthrough
 	case strings.Contains(osImage, "debian"):
-		queryPackagesCmd = "dpkg"
-		options = append(options, "-l")
-		packages = append(packages, "nfs-common", "open-iscsi", "cryptsetup", "dmsetup")
-		pipeFlag = true
+		packageProbeExecutables["nfs-common"] = "mount.nfs"
```
Comment (COLDTURNIP) [inline controller/monitor/environment_check_monitor.go:191]: We don't directly use mount.nfs in Longhorn. Just ensure the system NFS capability is fulfilled.

File: controller/monitor/environment_check_monitor.go [hunk 3]
```diff
@@ -233,36 +235,27 @@ func (m *EnvironmentCheckMonitor) syncPackagesInstalled(kubeNode *corev1.Node, n
...
-		if !pipeFlag {
-			args = append(args, pkg)
-		}
-		queryResult, err := nsexec.Execute(nil, queryPackagesCmd, args, lhtypes.ExecuteDefaultTimeout)
-		if err != nil {
+	installedPackages := []string{}
+	notInstalledPackages := []string{}
+	for pkg, executable := range packageProbeExecutables {
+		shellCmd := fmt.Sprintf("command -v %s", executable)
+		if _, err := nsexec.Execute(nil, "sh", []string{"-c", shellCmd}, lhtypes.ExecuteDefaultTimeout); err != nil {
```
Comment (COLDTURNIP) [inline controller/monitor/environment_check_monitor.go:242]: `command` is a shell function, not an independent executable, hence we have to wrap with `sh -c "..."`. The reason we don't use `which` is to minimize dependency. `which` is not installed on some Linux distro by default.

**Review by derekbit — APPROVED** [review]
LGTM

**Comment by derekbit** [comment]
@mergify backport v1.7.x

**Comment by derekbit** [comment]
@mergify backport v1.8.x

File: controller/monitor/environment_check_monitor.go [hunk 4]
```diff
@@ -203,66 +201,49 @@ func (m *EnvironmentCheckMonitor) syncPackagesInstalled(kubeNode *corev1.Node, n
...
-		packages = append(packages, "net-fs/nfs-utils", "sys-block/open-iscsi", "sys-fs/cryptsetup", "sys-fs/lvm2")
+		packageProbeExecutables["net-fs/nfs-utils"] = "mount.nfs"
+		packageProbeExecutables["sys-block/open-iscsi"] = "iscsiadm"
+		packageProbeExecutables["sys-fs/cryptsetup"] = "cryptsetup"
+		packageProbeExecutables["sys-fs/lvm2"] = "dmsetup"
 	default:
 		collectedData.conditions = types.SetCondition(collectedData.conditions, longhorn.NodeConditionTypeRequiredPackages, longhorn.ConditionStatusFalse,
 			string(longhorn.NodeConditionReasonUnknownOS),
 			fmt.Sprintf("Unable to verify the required packages because the OS image '%v' is unknown to the Longhorn system. Please ensure the required packages are installed.", osImage))
 		return
```
Comment (derekbit) [inline controller/monitor/environment_check_monitor.go:227]: I'm wondering if we can remove the message if Longhorn doesn't support the OS.

**Review by mantissahz — APPROVED** [review]
LGTM

File: controller/monitor/environment_check_monitor.go (hunk 4 above)
Comment (derekbit) [reply controller/monitor/environment_check_monitor.go:227]: WDYT? @COLDTURNIP @mantissahz

Comment (mantissahz) [reply controller/monitor/environment_check_monitor.go:227]: Do we have another place/information to show users that the OS is unsupported except for official documents?
If users are aware of the unsupported OS and responsible for it when installing Longhorn, this message is annoying.

In general, I'm fine to remove it.

Comment (derekbit) [reply controller/monitor/environment_check_monitor.go:227]: > Do we have another place/information to show users that the OS is unsupported except for official documents?
https://longhorn.io/docs/1.8.0/best-practices/#operating-system

Yes, we don't need to handle the OSes that Longhorn doesn't support or handle.

Comment (COLDTURNIP) [reply controller/monitor/environment_check_monitor.go:227]: Agree to remove. Maybe we can show such warning on the UI, or we already have CLI preflight checker at least. Node condition is too deep for user to get notice.

Comment (derekbit) [reply controller/monitor/environment_check_monitor.go:227]: > Node condition is too deep for user to get notice.

Node conditions are good for troubleshooting and self-diagnosis. However, there are too many OSes in the market, so we only need to take care of the OSes mentioned in the official doc.

**Review by derekbit — DISMISSED** [review]

File: controller/monitor/environment_check_monitor.go (hunk 4 above)
Comment (COLDTURNIP) [reply controller/monitor/environment_check_monitor.go:227]: Will append a commit to remove this condition for unsupported OSes.

Comment (derekbit) [reply controller/monitor/environment_check_monitor.go:227]: Isn't it removed?

Comment (COLDTURNIP) [reply controller/monitor/environment_check_monitor.go]: Yeah, we're going to remove the condition completely on unsupported OSes.

**Review by derekbit — APPROVED** [review]
LGTM

**Review by mantissahz — APPROVED** [review]

**Pull request merged**